pkg/blockbuilder/blockbuilder.go

```diff
@@ -287,29 +372,50 @@ func (b *BlockBuilder) consumePartitionSection(
 	})
 	defer b.kafkaClient.RemoveConsumePartitions(map[string][]int32{b.cfg.Kafka.Topic: {partition}})
 
+	var fetchPoller fetchPoller = b.kafkaClient
+
+	if b.cfg.Kafka.FetchConcurrencyMax > 0 {
+		f, ferr := b.newFetchers(ctx, logger, partition, startOffset)
+		if ferr != nil {
+			return fmt.Errorf("creating concurrent fetcher: %w", ferr)
+		}
+
+		b.readerMetricsSource.set(f)
+
+		f.Start(ctx)
+		defer f.Stop()
+
+		fetchPoller = &fetchWrapper{f}
+	}
+
 	level.Info(logger).Log("msg", "start consuming", "partition", partition, "start_offset", startOffset, "end_offset", endOffset)
 
-	var (
-		firstRecOffset = int64(-1)
-		lastRecOffset  = int64(-1)
-	)
+	firstRecOffset := int64(-1)
+	lastRecOffset := int64(-1)
```

Comment:
(nit) the `lastRecOffset` should be replaced with the `lastConsumedOffset`, that is defined at the top. The latter is what the top-most defer logs to track the over-consumption.

This chunk of the log is nice-to-have, thus nit. We can fix it separately.